class.c

```diff
@@ -636,184 +636,264 @@ Perl_class_apply_attributes(pTHX_ HV *stash, OP *attrlist)
+                                 "%" HvNAMEf_QUOTEDPREFIX " constructor",
+                                 SVfARG(paramname), HvNAMEfARG(stash));
+                    valop = newLISTOPn(OP_DIE, 0,
+                                       newSVOP(OP_CONST, 0, message),
+                                       NULL);
+                }
+
+                OP *helemop =
+                    newBINOP(OP_HELEM, 0,
+                             newPADxVOP(OP_PADHV, OPf_REF, PADIX_PARAMS),
+                             newSVOP(OP_CONST, 0, SvREFCNT_inc(paramname)));
+
+                if(PadnameFIELDINFO(pn)->def_if_undef) {
+                    /* delete $params{$paramname} // DEFOP */
+                    valop = newLOGOP(OP_DOR, 0,
+                                     newUNOP(OP_DELETE, 0, helemop), valop);
+                }
+                else if(PadnameFIELDINFO(pn)->def_if_false) {
+                    /* delete $params{$paramname} || DEFOP */
+                    valop = newLOGOP(OP_OR, 0,
+                                     newUNOP(OP_DELETE, 0, helemop), valop);
+                }
+                else {
+                    /* exists $params{$paramname} ? delete $params{$paramname} : DEFOP */
+                    /* more efficient with the new OP_HELEMEXISTSOR */
+                    valop = newLOGOP(OP_HELEMEXISTSOR, OPpHELEMEXISTSOR_DELETE << 8,
+                                     helemop, valop);
+                }
+
+                valop = op_contextualize(valop, G_SCALAR);
             }
+            break;
+
+        case '@':
+            op_priv = OPpINITFIELD_AV;
+            break;
+
+        case '%':
+            op_priv = OPpINITFIELD_HV;
+            break;
 
-            ops = op_append_list(OP_LINESEQ, ops, fieldop);
+        default:
+            NOT_REACHED;
         }
 
-        /* initfields CV should not get class_wrap_method_body() called on its
-         * body. pretend it isn't a method for now */
-        CvIsMETHOD_off(PL_compcv);
-        CV *initfields = newATTRSUB(floor_ix, NULL, NULL, NULL, ops);
-        CvIsMETHOD_on(initfields);
+        UNOP_AUX_item *aux;
+        aux = (UNOP_AUX_item *)PerlMemShared_malloc(sizeof(UNOP_AUX_item) * 2);
 
-        aux->xhv_class_initfields_cv = initfields;
-    }
-    else {
-        /* we had errors, clean up and don't populate initfields */
-        PADNAMELIST *fieldnames = aux->xhv_class_fields;
-        if (fieldnames) {
-            for(SSize_t i = PadnamelistMAX(fieldnames); i >= 0 ; i--) {
-                PADNAME *pn = PadnamelistARRAY(fieldnames)[i];
-                op_free(PadnameFIELDINFO(pn)->defop);
-            }
+        aux[0].uv = fieldix;
+
+        OP *fieldop = newUNOP_AUX(OP_INITFIELD, valop ? OPf_STACKED : 0, valop, aux);
+        fieldop->op_private = op_priv;
+
+        HE *he;
+        if((he = hv_fetch_ent(fieldix_to_padix, sv_2mortal(newSVuv(fieldix)), 0, 0)) &&
+           SvOK(HeVAL(he))) {
+            fieldop->op_targ = SvUV(HeVAL(he));
```

Comment:
factor out ```HeVAL(he)```, get the SV*. its too many layers deep of derefs in code, and probably does optimize, but its 3 or 4 derefs now from the C auto. if accidents happen with multi eval, they will be smaller.

Reply:
This code was only re-indented